File: dnn-providers/hip-kernel-provider/rocKE/tests/instances/parity/gfx950_attention_tiled_2d_fastkv_regp_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/gfx950_attention_tiled_2d_fastkv_regp_emit.py -- Python reference
# emitter for the experimental gfx950 "fast paged-KV descriptor + register-P"
# tiled-2D unified-attention kernel parity harness.
#
# Selects one of the sampled configs by argv[1], builds a
# UnifiedAttention2DTiledSpec with the experiment knobs, emits the kernel via
# build_unified_attention_2d_fastkv_register_p(spec, arch="gfx950") and prints
# lower_kernel_to_llvm(kernel, arch="gfx950") to stdout so it can be
# byte-compared with the C emitter gfx950_attention_tiled_2d_fastkv_regp_emit.c.
from rocke.instances.gfx950.attention_tiled_2d import UnifiedAttention2DTiledSpec
from rocke.instances.gfx950.attention_tiled_2d_fastkv_regp import (
    build_unified_attention_2d_fastkv_register_p,
)
from _emit_common import run_emit


# Shared base for the bf16 d64_b32_h64kv8 / T=64 / num_warps=4 experiment family.
_BASE = dict(
    head_size=64,
    block_size=32,
    num_query_heads=64,
    num_kv_heads=8,
    dtype="bf16",
    use_sinks=False,
    sliding_window=0,
    has_softcap=False,
    num_warps=4,
    waves_per_eu=2,
    tile_size=64,
    block_m_per_warp=32,
    use_mfma_32x32=True,
    use_transposed_qk_32x32=True,
    use_transposed_scalar_state=True,
    use_transposed_mask_once=True,
    use_fast_paged_kv_desc=True,
)


_CONFIGS = {
    0: dict(_BASE),
    1: dict(_BASE, use_transposed_half_local_pv=True),
    2: dict(_BASE, use_mfma32_skip_legacy_qreg=True),
    3: dict(_BASE, use_transposed_half_local_pv=True, use_mfma32_skip_legacy_qreg=True),
    4: dict(_BASE, use_agpr_alloc_zero=True, use_transposed_half_local_pv=True),
    5: dict(_BASE, use_grouped_kv2_softmax=True),
}


def _kernel(idx: int):
    if idx not in _CONFIGS:
        raise SystemExit(f"unknown config index {idx}")
    spec = UnifiedAttention2DTiledSpec(**_CONFIGS[idx])
    return build_unified_attention_2d_fastkv_register_p(spec, arch="gfx950")


def _emit_build(kernel, arch=None):
    return kernel


def main() -> int:
    return run_emit(
        _kernel,
        _emit_build,
        usage="usage: gfx950_attention_tiled_2d_fastkv_regp_emit.py <config_index>\n",
    )


if __name__ == "__main__":
    raise SystemExit(main())

File: dnn-providers/hip-kernel-provider/rocke/library/tests/parity/gfx950_attention_tiled_2d_fastkv_regp_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/gfx950_attention_tiled_2d_fastkv_regp_emit.c -- C-side emitter
 * for the experimental gfx950 "fast paged-KV descriptor + register-P" tiled-2D
 * unified-attention parity harness.
 *
 * Selects one of the sampled configs by argv[1], fills a
 * rocke_attention_tiled_2d_spec_t identically to the Python emitter
 * gfx950_attention_tiled_2d_fastkv_regp_emit.py, builds the kernel via
 * rocke_build_unified_attention_2d_fastkv_register_p(&b, &spec, "gfx950"), lowers
 * it with rocke_lower_kernel_to_llvm(kernel, AUTO, "gfx950", ...) and prints the
 * .ll to stdout for byte comparison.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/instance_gfx950_attention_tiled_2d_fastkv_regp.h"
#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Fill `s` for config index `idx`. Returns 0 on success, -1 on unknown idx.
 *
 * Shared base for the bf16 d64_b32_h64kv8 / T=64 / num_warps=4 experiment
 * family, then per-config additive flags. Mirrors the Python emitter's _BASE /
 * _CONFIGS exactly. */
static void make_base(rocke_attention_tiled_2d_spec_t* s)
{
    *s = rocke_attention_tiled_2d_spec_default();
    s->head_size = 64;
    s->block_size = 32;
    s->num_query_heads = 64;
    s->num_kv_heads = 8;
    s->dtype = "bf16";
    s->use_sinks = false;
    s->sliding_window = 0;
    s->has_softcap = false;
    s->num_warps = 4;
    s->has_waves_per_eu = true;
    s->waves_per_eu = 2;
    s->has_tile_size = true;
    s->tile_size = 64;
    s->block_m_per_warp = 32;
    s->use_mfma_32x32 = true;
    s->use_transposed_qk_32x32 = true;
    s->use_transposed_scalar_state = true;
    s->use_transposed_mask_once = true;
    s->use_fast_paged_kv_desc = true;
}

static int make_spec(int idx, rocke_attention_tiled_2d_spec_t* s)
{
    make_base(s);
    switch(idx)
    {
    case 0:
        break;
    case 1:
        s->use_transposed_half_local_pv = true;
        break;
    case 2:
        s->use_mfma32_skip_legacy_qreg = true;
        break;
    case 3:
        s->use_transposed_half_local_pv = true;
        s->use_mfma32_skip_legacy_qreg = true;
        break;
    case 4:
        s->use_agpr_alloc_zero = true;
        s->use_transposed_half_local_pv = true;
        break;
    case 5:
        s->use_grouped_kv2_softmax = true;
        break;
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(stderr, "usage: %s <config_index>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    rocke_attention_tiled_2d_spec_t s;
    if(make_spec(idx, &s) != 0)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 1;
    }

    rocke_ir_builder_t b;
    if(rocke_ir_builder_init(&b, "attention_tiled_2d_fastkv_regp") != ROCKE_OK)
    {
        fprintf(stderr, "builder init failed\n");
        return 1;
    }

    rocke_kernel_def_t* kernel
        = rocke_build_unified_attention_2d_fastkv_register_p(&b, &s, "gfx950");
    if(!kernel)
    {
        fprintf(stderr, "build failed: err=%s\n", rocke_ir_builder_error(&b));
        rocke_ir_builder_free(&b);
        return 1;
    }

    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        rocke_status_t st
            = rocke_lower_kernel_to_llvm(kernel, ROCKE_LLVM_FLAVOR_AUTO, "gfx950", &llvm_text);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* t = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &t);
        if(st != ROCKE_OK || !t)
        {
            fprintf(stderr, "serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(t, stdout);
        free(t);
    }
    else if(strcmp(mode, "verify") == 0)
    {
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s2 = rocke_diag_to_string(&d[i]);
            if(s2)
            {
                puts(s2);
                free(s2);
            }
        }
        rocke_diags_free(d, n);
    }
    else
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        rocke_ir_builder_free(&b);
        return 2;
    }

    rocke_ir_builder_free(&b);
    return 0;
}
